File: sdk/python/neura/webhook.py
"""Webhook operations — register, list, manage webhook endpoints."""
from typing import Any, Optional
from .client import HttpClient


class WebhookAPI:
    """High-level webhook operations."""

    def __init__(self, client: HttpClient):
        self._client = client

    def create(self, url: str, events: list[str], secret: Optional[str] = None) -> dict:
        """Register a webhook for event notifications."""
        body = {"url": url, "events": events}
        if secret:
            body["secret"] = secret
        result = self._client.request("POST", "/api/webhooks", body)
        return result["data"]

    def list(self) -> list[dict]:
        """List all registered webhooks."""
        result = self._client.request("GET", "/api/webhooks")
        return result["data"]

    def get(self, webhook_id: str) -> dict:
        """Get webhook details by ID."""
        result = self._client.request("GET", f"/api/webhooks/{webhook_id}")
        return result["data"]

    def update(self, webhook_id: str, **kwargs) -> dict:
        """Update a webhook's url, events, or secret."""
        result = self._client.request("PATCH", f"/api/webhooks/{webhook_id}", kwargs)
        return result["data"]

    def delete(self, webhook_id: str) -> None:
        """Delete a webhook."""
        self._client.request("DELETE", f"/api/webhooks/{webhook_id}")

    def retry_failed(self) -> dict:
        """Trigger retry of all failed webhook deliveries due for retry."""
        result = self._client.request("POST", "/api/webhooks/retry")
        return result["data"]

File: sdk/typescript/src/webhook.ts
// ============================================================
// Neura SDK — Webhook Operations
// ============================================================

import { HttpClient } from './client'
import type { Webhook, CreateWebhookInput } from './types'

export class WebhookAPI {
  constructor(private client: HttpClient) {}

  /** Register a webhook for event notifications. */
  async create(input: CreateWebhookInput): Promise<Webhook> {
    const { data } = await this.client.request<Webhook>('POST', '/api/webhooks', input)
    return data
  }

  /** List all registered webhooks. */
  async list(): Promise<Webhook[]> {
    const { data } = await this.client.request<Webhook[]>('GET', '/api/webhooks')
    return data
  }

  /** Get webhook details by ID. */
  async get(id: string): Promise<Webhook> {
    const { data } = await this.client.request<Webhook>('GET', `/api/webhooks/${id}`)
    return data
  }

  /** Update a webhook. */
  async update(id: string, input: Partial<CreateWebhookInput>): Promise<Webhook> {
    const { data } = await this.client.request<Webhook>('PATCH', `/api/webhooks/${id}`, input)
    return data
  }

  /** Delete a webhook. */
  async delete(id: string): Promise<void> {
    await this.client.request('DELETE', `/api/webhooks/${id}`)
  }

  /** Trigger retry of all failed webhook deliveries due for retry. */
  async retryFailed(): Promise<{ processed: number; succeeded: number; failed: number }> {
    const { data } = await this.client.request<{ processed: number; succeeded: number; failed: number }>(
      'POST', '/api/webhooks/retry'
    )
    return data
  }
}
